### ISSUE TITLE
org/arangodb/crypto support for public key algorithms

### ISSUE THREAD

**Issue by bangert** [arangodb/arangodb#1786]
Hi,

a foxx app, wanting to verify JWTs signed with the RS256 algorithm, needs to do public key crypto in arangodb.

Specifically, we use Keycloak as the authentication provider, and it can only provide private key signed JWTs.

See also https://issues.jboss.org/browse/KEYCLOAK-1204

Thanks

**Comment by pluma** [comment]
I would love to see more crypto, too. Not sure how much the libraries we bundle / depend on support cross-platform at the moment, though.

**Comment by dothebart** [comment]
We're currently using OpenSSL to provide crytpographic functions. i.e. HMAC is mentioned in this ticket: https://www.openssl.org/docs/manmaster/crypto/hmac.html
Is well provided. 

I can't seem to find anything named RS256, only AEL256 https://openssl.org/docs/manmaster/apps/rsa.html Is this what you refer to? If not, can you please elaborate which parts of openssl or other c/c++ crypto libraries implement this? Since [we're not intending to enter the crypto business](http://stackoverflow.com/questions/1914257/when-can-you-trust-yourself-to-implement-cryptography-based-solutions) by ourselves ;-)

**Comment by bangert** [comment]
Good call on not entering the crypto business yourselves ;-)

RS256 uses RSA crypto instead of HMAC.
So instead of a shared secret you use the private key at the identity provider and the public key in the application. I otherwords openssl has everything necessary.

https://github.com/brianloveswords/node-jwa is an implementation of these algorithms base on nodes crypto module.

Looking at node's crypto module, i would like functionality equivalent of the Verify class
https://nodejs.org/api/crypto.html#crypto_class_verify

ES256 is then using elliptic curve crypto, and i guess would need the ECDH class.

For more info on the different algorithm types used in JWT
https://tools.ietf.org/html/rfc7518

**Comment by dothebart** [comment]
Related:  #1771 Requires a C++ implementation similar to nodes node_crypto/SignBase class. 
Probably not achieveable for the 3.0x release series.

**Comment by bangert** [comment]
From https://issues.jboss.org/browse/KEYCLOAK-1204:
"The most commonly used way to sign JWTs are RS256 so ArangoDB should really support that."

Any news for the roadmap?
Thanks

**Comment by pluma** [comment]
@dothebart the RSA docs you linked to have moved: https://www.openssl.org/docs/manmaster/man1/rsa.html

But yes, OpenSSL seems to support RSA, so it should be possible to add support for this. Supporting public/private key encryption would probably be useful outside of JWT too.

**Comment by Calrain** [comment]
[RS256](http://connect2id.com/products/nimbus-jose-jwt/examples/jwt-with-rsa-signature) refers to the RSA implementation of SHA256, which is described as `RSA PKCS#1 signature with SHA-256`

When trying to verify a JWT bearer token from Azure Active Directory (AAD) with `crypto.jwtDecode` you get this error:

    Unknown algorithm "RS256". Only the following algorithms are supported at this time: HS256, HS384, HS512, none
Note: this is using `const crypto = require('@arangodb/crypto')`


If you manually decrypt the header of an AAD JWT bearer token it looks like this:

    "header":{
      "typ":"JWT",
      "alg":"RS256",
      "x5t":"RrQ...GRM",
      "kid":"RrQ...GRM"
    }

Foxx supports the `sha256` algorithm, but the HMAC version, not the RSA version (which is required for RS256). This was validated by looking inside `@arangodb/crypto.js`, it shows that HMAC is being used for the `sha256` hash.

There is an article on [StackOverflow with people having issues validating RS256 tokens in C#](http://stackoverflow.com/questions/34403823/verifying-jwt-signed-with-the-rs256-algorithm-using-public-key-in-c-sharp) and they use the `RSACryptoServiceProvider`, but that is C#.

So, there are two versions of sha256, HMAC SHA256 and RSA SHA256, it would be great to get the RSA version added to Foxx.

Unfortunately I'm out of my depth to help more on this, other than test!

**Comment by pluma** [comment]
@Calrain SHA256 is a hashing algorithm. RSA is a public key encryption algorithm. HMAC is a key derivation algorithm. Both can use different hashing algorithms, like SHA256.

There's currently no support for any encryption algorithms in ArangoDB but we are using OpenSSL, which seems to support RSA (and more), so it mostly comes down to wrapping the existing OpenSSL methods in C++ and exposing them to JavaScript.

However because RSA is an actual encryption algorithm it's not quite as straightforward as wrapping HMAC. We also need the methods for generating public/private key pairs. I believe there are also native methods for signing and verifying signatures.

Specifically someone would need to do the following:

1. Evaluate the libraries we bundle (specifically OpenSSL) and whether they support everything we need, cross-platform (this is the critical part).

2. Wrap the relevant methods in some V8-friendly C++ functions and expose them in JS.

3. Wrap the JS methods in the crypto module and document them.

**Comment by Calrain** [comment]
@pluma Thanks for the clarification, it does seem to be more work than the HMAC implementation.

**Comment by innwerkler** [comment]
Just as a hint, I managed work around this and implement RS256 token verification with the node-forge module (https://github.com/digitalbazaar/forge), though this required two minimal patches in the source code.

**Comment by u382514** [comment]
@innwerkler Can you give us more than a hint on how you did it and what patches in what source code are you referring to?

**Comment by innwerkler** [comment]
Of course. The problem I fell over was that node-forge has a check to figure out whether it is running in Node.js, which did not really work in Foxx:

_// check if running under Node.js
util.isNodejs =
  typeof process !== 'undefined' && process.versions && process.versions.node;__

Initially I patched that line, but in the end I figured out that all seems to work if I just create process.versions and process.versions.node before the import:

process.versions = {
  node: '42',
  http_parser: '',
  v8: '',
  ares: '',
  uv: '',
  zlib: '',
  modules: '',
  openssl: ''
};
//     "node-forge": "^1.3.1",
const forge = require('node-forge');

Probably not the cleanest approach, but it helped me to solve the problem at hand at least.
